paasta_tools/utils.py

```diff
@@ -1015,6 +1015,11 @@ def get_deploy_whitelist(self):
         """
         return self.get("deploy_whitelist", [])
 
+    def get_expected_slave_attributes(self):
+        """Return a list of dictionaries, representing the expected combinations of attributes in this cluster. Used for
```

Comment:
The list of dictionaries (each representing an example slave) helps us when we have constraints. E.g. if you `discover: habitat` but have `deploy_whitelist: [region, [uswest1-prod]]`, then we need to set constraints `['habitat', 'GROUP_BY', '2']` instead of `['habitat', 'GROUP_BY', '3']`

I could also figure some of this out by reading /nail/etc/services/location_mapping.json instead of a new configuration key, but that doesn't have any information about e.g. pools -- if you use a pool that only exists in one region, you've effectively added a `[region, LIKE, ...]` constraint